src/ray/raylet/node_manager.cc

```diff
@@ -2766,6 +2769,91 @@ void NodeManager::SendSpilledObjectRestorationRequestToRemoteNode(
       });
 }
 
+void NodeManager::FillNormalTaskResourceUsage(
+    const std::shared_ptr<rpc::ResourcesData> &resources_data) {
```

Comment:
This is going to be the only part of the heartbeat that we can't unit test. Can we instead make this a utility function that we can write unit tests for (then write the tests)?

The new signature might look something like:

```
void FillNormalTaskResourceUsage(const std::shared_ptr<rpc::ResourcesData> &resources_data, const std::unordered_map<WorkerID, std::shared_ptr<WorkerInterface>> &leased_workers, const std::shared_ptr<std::unordered_map<std::string, double>> &last_report_normal_resources_)
```

Reply:
Hey @wuisawesome , `node_manager` has no unit test at present. Do we need to add a `node_manager_test` first?

Reply:
Yeah, or we can make it part of `ClusterTaskManagerInterface` and put it in `cluster_task_manager_test`